File: image_url_generator.js
/**
 * Image URL Generator - JavaScript/Node.js Version
 * Provides multiple methods to generate URLs for viewing images in web applications
 */

const fs = require('fs');
const path = require('path');
const http = require('http');
const express = require('express');

class ImageURLGenerator {
    constructor(baseUrl = 'http://localhost:3000') {
        this.baseUrl = baseUrl;
        this.server = null;
    }

    /**
     * Generate a file URL for local images (browser file:// protocol)
     */
    fileUrl(imagePath) {
        const absolutePath = path.resolve(imagePath);
        return `file://${absolutePath}`;
    }

    /**
     * Generate a data URL (base64 encoded) for images
     */
    dataUrl(imagePath) {
        if (!fs.existsSync(imagePath)) {
            throw new Error(`Image not found: ${imagePath}`);
        }

        const imageBuffer = fs.readFileSync(imagePath);
        const mimeType = this.getMimeType(imagePath);
        const base64String = imageBuffer.toString('base64');
        
        return `data:${mimeType};base64,${base64String}`;
    }

    /**
     * Generate a web URL for images served by a web server
     */
    webUrl(imagePath, relativeToWebroot = null) {
        let relativePath;
        
        if (relativeToWebroot) {
            relativePath = path.relative(relativeToWebroot, imagePath);
        } else {
            relativePath = path.basename(imagePath);
        }
        
        return new URL(relativePath, this.baseUrl + '/').href;
    }

    /**
     * Get MIME type for image files
     */
    getMimeType(imagePath) {
        const ext = path.extname(imagePath).toLowerCase();
        const mimeTypes = {
            '.jpg': 'image/jpeg',
            '.jpeg': 'image/jpeg',
            '.png': 'image/png',
            '.gif': 'image/gif',
            '.bmp': 'image/bmp',
            '.webp': 'image/webp',
            '.svg': 'image/svg+xml'
        };
        return mimeTypes[ext] || 'image/jpeg';
    }

    /**
     * Start a local Express server to serve images
     */
    async startLocalServer(directory = '.', port = 3000) {
        const express = require('express');
        const app = express();
        
        // Serve static files from the specified directory
        app.use(express.static(directory));
        
        // API endpoint to get image URLs
        app.get('/api/image-urls/:filename', (req, res) => {
            const filename = req.params.filename;
            const imagePath = path.join(directory, filename);
            
            if (!fs.existsSync(imagePath)) {
                return res.status(404).json({ error: 'Image not found' });
            }
            
            try {
                const urls = this.generateAllUrls(imagePath);
                res.json(urls);
            } catch (error) {
                res.status(500).json({ error: error.message });
            }
        });
        
        return new Promise((resolve, reject) => {
            this.server = app.listen(port, 'localhost', () => {
                this.baseUrl = `http://localhost:${port}`;
                console.log(`Local server started at ${this.baseUrl}`);
                resolve(this.baseUrl);
            });
            
            this.server.on('error', reject);
        });
    }

    /**
     * Stop the local server
     */
    stopLocalServer() {
        if (this.server) {
            this.server.close(() => {
                console.log('Local server stopped');
            });
            this.server = null;
        }
    }

    /**
     * Generate all possible URL formats for an image
     */
    generateAllUrls(imagePath) {
        return {
            fileUrl: this.fileUrl(imagePath),
            dataUrl: this.dataUrl(imagePath),
            webUrl: this.webUrl(imagePath)
        };
    }

    /**
     * Create an HTML viewer for the image with all URL types
     */
    createHtmlViewer(imagePath, outputFile = 'image_viewer.html') {
        const urls = this.generateAllUrls(imagePath);
        const imageName = path.basename(imagePath);
        
        const htmlContent = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Image Viewer - ${imageName}</title>
    <style>
        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            max-width: 1200px;
            margin: 0 auto;
            padding: 20px;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
        }
        .container {
            background: white;
            border-radius: 15px;
            padding: 30px;
            box-shadow: 0 10px 30px rgba(0,0,0,0.3);
            backdrop-filter: blur(10px);
        }
        .header {
            text-align: center;
            margin-bottom: 30px;
        }
        .header h1 {
            color: #333;
            margin-bottom: 10px;
        }
        .image-display {
            text-align: center;
            margin: 30px 0;
            padding: 20px;
            background: #f8f9fa;
            border-radius: 10px;
        }
        .image-display img {
            max-width: 100%;
            max-height: 400px;
            border-radius: 10px;
            box-shadow: 0 8px 16px rgba(0,0,0,0.2);
            transition: transform 0.3s ease;
        }
        .image-display img:hover {
            transform: scale(1.05);
        }
        .url-section {
            margin: 20px 0;
            padding: 20px;
            background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%);
            border-radius: 10px;
            border-left: 5px solid #007bff;
            transition: all 0.3s ease;
        }
        .url-section:hover {
            transform: translateY(-2px);
            box-shadow: 0 5px 15px rgba(0,0,0,0.1);
        }
        .url-type {
            font-weight: bold;
            color: #007bff;
            margin-bottom: 10px;
            font-size: 16px;
        }
        .url-description {
            color: #666;
            font-size: 14px;
            margin-bottom: 10px;
        }
        .url-text {
            font-family: 'Courier New', monospace;
            background: #fff;
            padding: 12px;
            border-radius: 5px;
            word-break: break-all;
            font-size: 12px;
            border: 1px solid #e9ecef;
            max-height: 100px;
            overflow-y: auto;
        }
        .copy-btn {
            background: linear-gradient(135deg, #007bff 0%, #0056b3 100%);
            color: white;
            border: none;
            padding: 8px 15px;
            border-radius: 5px;
            cursor: pointer;
            font-size: 12px;
            margin-top: 10px;
            transition: all 0.3s ease;
        }
        .copy-btn:hover {
            transform: translateY(-1px);
            box-shadow: 0 3px 10px rgba(0,123,255,0.3);
        }
        .success-message {
            color: #28a745;
            font-size: 12px;
            margin-top: 5px;
            opacity: 0;
            transition: opacity 0.3s ease;
        }
        .success-message.show {
            opacity: 1;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🖼️ Image Viewer</h1>
            <p><strong>${imageName}</strong></p>
        </div>
        
        <div class="image-display">
            <img src="${urls.dataUrl}" alt="${imageName}" loading="lazy">
        </div>
        
        <div class="url-section">
            <div class="url-type">📁 File URL</div>
            <div class="url-description">Direct filesystem access (works locally)</div>
            <div class="url-text">${urls.fileUrl}</div>
            <button class="copy-btn" onclick="copyToClipboard('${urls.fileUrl.replace(/'/g, "\\'")}', this)">Copy URL</button>
            <div class="success-message">Copied to clipboard!</div>
        </div>
        
        <div class="url-section">
            <div class="url-type">🌐 Web URL</div>
            <div class="url-description">HTTP server access (shareable link)</div>
            <div class="url-text">${urls.webUrl}</div>
            <button class="copy-btn" onclick="copyToClipboard('${urls.webUrl}', this)">Copy URL</button>
            <div class="success-message">Copied to clipboard!</div>
        </div>
        
        <div class="url-section">
            <div class="url-type">📊 Data URL</div>
            <div class="url-description">Base64 encoded (embedded, works anywhere)</div>
            <div class="url-text">${urls.dataUrl.substring(0, 100)}...</div>
            <button class="copy-btn" onclick="copyToClipboard('${urls.dataUrl.replace(/'/g, "\\'")}', this)">Copy Full Data URL</button>
            <div class="success-message">Copied to clipboard!</div>
        </div>
    </div>
    
    <script>
        async function copyToClipboard(text, button) {
            try {
                await navigator.clipboard.writeText(text);
                const successMsg = button.nextElementSibling;
                successMsg.classList.add('show');
                setTimeout(() => {
                    successMsg.classList.remove('show');
                }, 2000);
            } catch (err) {
                console.error('Could not copy text: ', err);
                alert('Failed to copy to clipboard');
            }
        }
        
        // Add some interactivity
        document.addEventListener('DOMContentLoaded', function() {
            const urlSections = document.querySelectorAll('.url-section');
            urlSections.forEach((section, index) => {
                section.style.animationDelay = index * 0.1 + 's';
            });
        });
    </script>
</body>
</html>`;
        
        fs.writeFileSync(outputFile, htmlContent);
        return path.resolve(outputFile);
    }
}

/**
 * Browser-compatible version for client-side usage
 */
class BrowserImageURLGenerator {
    constructor(baseUrl = window.location.origin) {
        this.baseUrl = baseUrl;
    }

    /**
     * Generate a blob URL for File objects
     */
    blobUrl(file) {
        return URL.createObjectURL(file);
    }

    /**
     * Generate data URL from File object
     */
    async dataUrlFromFile(file) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.onload = () => resolve(reader.result);
            reader.onerror = reject;
            reader.readAsDataURL(file);
        });
    }

    /**
     * Generate web URL for uploaded files
     */
    webUrl(filename) {
        return new URL(`/uploads/${filename}`, this.baseUrl).href;
    }

    /**
     * Create image preview with URLs
     */
    async createImagePreview(file, container) {
        const dataUrl = await this.dataUrlFromFile(file);
        const blobUrl = this.blobUrl(file);
        
        const previewHtml = `
            <div class="image-preview">
                <img src="${dataUrl}" alt="${file.name}" style="max-width: 300px; max-height: 200px;">
                <div class="url-info">
                    <p><strong>Blob URL:</strong> <code>${blobUrl}</code></p>
                    <p><strong>Data URL:</strong> <code>${dataUrl.substring(0, 50)}...</code></p>
                </div>
            </div>
        `;
        
        if (typeof container === 'string') {
            document.getElementById(container).innerHTML = previewHtml;
        } else {
            container.innerHTML = previewHtml;
        }
        
        return { dataUrl, blobUrl };
    }
}

// Demo function for Node.js
async function demo() {
    const generator = new ImageURLGenerator();
    
    console.log('Image URL Generator Demo (JavaScript)');
    console.log('=' .repeat(40));
    
    // For demo, we'll assume there's an image file
    const imagePath = process.argv[2] || 'demo_image.jpg';
    
    if (!fs.existsSync(imagePath)) {
        console.log(`Error: Image not found at ${imagePath}`);
        console.log('Please provide an image path as an argument.');
        return;
    }
    
    console.log(`Generating URLs for: ${imagePath}`);
    console.log('-'.repeat(40));
    
    try {
        const urls = generator.generateAllUrls(imagePath);
        
        console.log(`File URL: ${urls.fileUrl}`);
        console.log(`Web URL: ${urls.webUrl}`);
        console.log(`Data URL: ${urls.dataUrl.substring(0, 80)}...`);
        
        // Create HTML viewer
        const htmlFile = generator.createHtmlViewer(imagePath);
        console.log(`Created HTML viewer: ${htmlFile}`);
        
        // Start server
        const serverUrl = await generator.startLocalServer();
        console.log(`Server started at: ${serverUrl}`);
        console.log(`View image at: ${serverUrl}/image_viewer.html`);
        
        // Keep server running
        process.on('SIGINT', () => {
            generator.stopLocalServer();
            process.exit(0);
        });
        
    } catch (error) {
        console.error(`Error: ${error.message}`);
    }
}

// Export for both Node.js and browser environments
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { ImageURLGenerator, BrowserImageURLGenerator };
    
    // Run demo if this file is executed directly
    if (require.main === module) {
        demo();
    }
} else {
    // Browser environment
    window.ImageURLGenerator = BrowserImageURLGenerator;
}

File: image_url_generator.py
#!/usr/bin/env python3
"""
Image URL Generator
Provides multiple methods to generate URLs for viewing images
"""

import os
import base64
import mimetypes
from pathlib import Path
from urllib.parse import urljoin
from http.server import HTTPServer, SimpleHTTPRequestHandler
import threading
import webbrowser

class ImageURLGenerator:
    def __init__(self, base_url="http://localhost:8000"):
        self.base_url = base_url
        self.server = None
        self.server_thread = None
    
    def file_url(self, image_path):
        """Generate a file:// URL for local images"""
        abs_path = os.path.abspath(image_path)
        return f"file://{abs_path}"
    
    def data_url(self, image_path):
        """Generate a data URL (base64 encoded) for images"""
        if not os.path.exists(image_path):
            raise FileNotFoundError(f"Image not found: {image_path}")
        
        # Get MIME type
        mime_type, _ = mimetypes.guess_type(image_path)
        if not mime_type or not mime_type.startswith('image/'):
            mime_type = 'image/jpeg'  # Default fallback
        
        # Read and encode image
        with open(image_path, 'rb') as image_file:
            encoded_string = base64.b64encode(image_file.read()).decode('utf-8')
        
        return f"data:{mime_type};base64,{encoded_string}"
    
    def web_url(self, image_path, relative_to_webroot=None):
        """Generate a web URL assuming the image is served by a web server"""
        if relative_to_webroot:
            rel_path = os.path.relpath(image_path, relative_to_webroot)
        else:
            rel_path = os.path.basename(image_path)
        
        return urljoin(self.base_url + "/", rel_path)
    
    def start_local_server(self, directory=".", port=8000):
        """Start a local HTTP server to serve images"""
        os.chdir(directory)
        
        class QuietHTTPRequestHandler(SimpleHTTPRequestHandler):
            def log_message(self, format, *args):
                pass  # Suppress log messages
        
        self.server = HTTPServer(('localhost', port), QuietHTTPRequestHandler)
        self.server_thread = threading.Thread(target=self.server.serve_forever)
        self.server_thread.daemon = True
        self.server_thread.start()
        
        self.base_url = f"http://localhost:{port}"
        print(f"Local server started at {self.base_url}")
        return self.base_url
    
    def stop_local_server(self):
        """Stop the local HTTP server"""
        if self.server:
            self.server.shutdown()
            self.server = None
            print("Local server stopped")
    
    def generate_all_urls(self, image_path):
        """Generate all possible URL formats for an image"""
        urls = {
            'file_url': self.file_url(image_path),
            'data_url': self.data_url(image_path),
            'web_url': self.web_url(image_path)
        }
        return urls
    
    def create_html_viewer(self, image_path, output_file="image_viewer.html"):
        """Create an HTML file to view the image with all URL types"""
        urls = self.generate_all_urls(image_path)
        
        html_content = f"""<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Image Viewer - {os.path.basename(image_path)}</title>
    <style>
        body {{
            font-family: Arial, sans-serif;
            max-width: 1200px;
            margin: 0 auto;
            padding: 20px;
            background-color: #f5f5f5;
        }}
        .container {{
            background: white;
            border-radius: 8px;
            padding: 30px;
            box-shadow: 0 2px 10px rgba(0,0,0,0.1);
        }}
        .image-display {{
            text-align: center;
            margin: 30px 0;
        }}
        .image-display img {{
            max-width: 100%;
            max-height: 500px;
            border-radius: 8px;
            box-shadow: 0 4px 8px rgba(0,0,0,0.1);
        }}
        .url-section {{
            margin: 20px 0;
            padding: 15px;
            background: #f8f9fa;
            border-radius: 5px;
            border-left: 4px solid #007bff;
        }}
        .url-type {{
            font-weight: bold;
            color: #007bff;
            margin-bottom: 8px;
        }}
        .url-text {{
            font-family: monospace;
            background: #e9ecef;
            padding: 8px;
            border-radius: 3px;
            word-break: break-all;
            font-size: 12px;
        }}
        .copy-btn {{
            background: #007bff;
            color: white;
            border: none;
            padding: 5px 10px;
            border-radius: 3px;
            cursor: pointer;
            font-size: 12px;
            margin-top: 5px;
        }}
        .copy-btn:hover {{
            background: #0056b3;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Image Viewer: {os.path.basename(image_path)}</h1>
        
        <div class="image-display">
            <img src="{urls['data_url']}" alt="{os.path.basename(image_path)}">
        </div>
        
        <div class="url-section">
            <div class="url-type">File URL (Local filesystem access)</div>
            <div class="url-text">{urls['file_url']}</div>
            <button class="copy-btn" onclick="copyToClipboard('{urls['file_url']}')">Copy URL</button>
        </div>
        
        <div class="url-section">
            <div class="url-type">Web URL (HTTP server access)</div>
            <div class="url-text">{urls['web_url']}</div>
            <button class="copy-btn" onclick="copyToClipboard('{urls['web_url']}')">Copy URL</button>
        </div>
        
        <div class="url-section">
            <div class="url-type">Data URL (Base64 encoded - works anywhere)</div>
            <div class="url-text">{urls['data_url'][:100]}...</div>
            <button class="copy-btn" onclick="copyToClipboard('{urls['data_url']}')">Copy Full Data URL</button>
        </div>
    </div>
    
    <script>
        function copyToClipboard(text) {{
            navigator.clipboard.writeText(text).then(function() {{
                alert('URL copied to clipboard!');
            }}, function(err) {{
                console.error('Could not copy text: ', err);
            }});
        }}
    </script>
</body>
</html>"""
        
        with open(output_file, 'w') as f:
            f.write(html_content)
        
        return os.path.abspath(output_file)

def main():
    """Demo usage of the ImageURLGenerator"""
    generator = ImageURLGenerator()
    
    print("Image URL Generator Demo")
    print("=" * 40)
    
    # Example usage
    image_path = input("Enter image path (or press Enter for demo): ").strip()
    
    if not image_path:
        # Create a demo image for testing
        print("Creating a demo image...")
        import io
        from PIL import Image, ImageDraw
        
        # Create a simple demo image
        img = Image.new('RGB', (400, 300), color='lightblue')
        draw = ImageDraw.Draw(img)
        draw.text((150, 140), "Demo Image", fill='black')
        img.save('demo_image.jpg')
        image_path = 'demo_image.jpg'
        print(f"Created demo image: {image_path}")
    
    if not os.path.exists(image_path):
        print(f"Error: Image not found at {image_path}")
        return
    
    print(f"\nGenerating URLs for: {image_path}")
    print("-" * 40)
    
    try:
        # Generate all URL types
        urls = generator.generate_all_urls(image_path)
        
        print(f"File URL: {urls['file_url']}")
        print(f"Web URL: {urls['web_url']}")
        print(f"Data URL: {urls['data_url'][:80]}...")
        
        # Start local server
        print("\nStarting local web server...")
        server_url = generator.start_local_server()
        
        # Create HTML viewer
        html_file = generator.create_html_viewer(image_path)
        print(f"Created HTML viewer: {html_file}")
        
        # Open in browser
        viewer_url = f"{server_url}/image_viewer.html"
        print(f"Opening viewer at: {viewer_url}")
        webbrowser.open(viewer_url)
        
        input("Press Enter to stop the server...")
        generator.stop_local_server()
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()